Avoid false removals in ClusterReceptionist, #26284

patriknw opened the pull request:
* The scenario was (probably) that a node was restarted with
  same host:port and then didn't join the same cluster. The DData
  Replicator in the original cluster would continue sending messages
  to the new incarnation resulting in false removals.
* The fix is that DData Replicator includes the system uid of the intended
  target in messages and if recipient gets a message that is not intended for
  its uid it will discard it and thereby not spreading information across
  different clusters.
* Reproduced in ClusterReceptionistSpec
* Much hardening of other things in ClusterReceptionistSpec
* There are also some improvements to ClusterReceptionist to not leak
  Listing with refs of removed nodes. Also an attempt to detect registered
  refs that have been removed from ddata. First thought was to re-add those
  but I don't think that should ever happen given the fix in ddata. That part
  can probably be removed, but could be good for logging purposes when
  testing this.

Refs #26284

patriknw committed:
Avoid false removals in ClusterReceptionist, #26284

* The scenario was (probably) that a node was restarted with
  same host:port and then didn't join the same cluster. The DData
  Replicator in the original cluster would continue sending messages
  to the new incarnation resulting in false removals.
* The fix is that DData Replicator includes the system uid of the sending
  or target system in messages and if recipient gets a message that is from/to
  unknown it will discard it and thereby not spreading information across
  different clusters.
* Reproduced in ClusterReceptionistSpec
* Much hardening of other things in ClusterReceptionistSpec
* There are also some improvements to ClusterReceptionist to not leak
  Listing with refs of removed nodes.
* use ClusterShuttingDown
* The reason for using sender system uid instead of target uid in messages
  like Read and Write is that then the optimization for sending same message
  to many destinations can remain.

patriknw commented on akka-cluster-typed/src/main/scala/akka/cluster/typed/internal/receptionist/Registry.scala:
```diff
@@ -23,23 +23,32 @@ import scala.concurrent.duration.Deadline
+ * @param registeredServices Local actors that were registered and not terminated. Used to find entries
+ *                           that have been removed from ddata but are know to be existing and should be
+ *                           re-added. Can happen when talking to a node outside the cluster such as
+ *                           a restarted node with same hostname and port.
```
the `registeredServices` is the part that I probably will remove again, but it can be useful for logging when testing this

johanandren commented on akka-cluster-typed/src/main/scala/akka/cluster/typed/internal/receptionist/ClusterReceptionist.scala:
```diff
@@ -247,63 +271,107 @@ private[typed] object ClusterReceptionist extends ReceptionistBehaviorProvider {
+                  serviceKey,
+                  newRegistry.activeActorRefsFor(serviceKey, selfUniqueAddress))
+                subscribers.foreach(_ ! listing)
+              }
```
The old logic would filter out tombstoned before notifying to avoid sending out a `Listing` with a removed local ref in the case of concurrent update. While this doesn't, wasn't that a good idea?

patriknw replied:
It still does that in `activeActorRefsFor`. Previous impl only did it for subscription notifications but not for the reply to Register and Find

johanandren reviewed (approved):
LGTM

patriknw commented on akka-cluster-typed/src/main/scala/akka/cluster/typed/internal/receptionist/ClusterReceptionist.scala:
```diff
@@ -39,15 +44,18 @@ private[typed] object ClusterReceptionist extends ReceptionistBehaviorProvider {
       else UniqueAddress(ref.path.address, systemUid)
-    override def toString = ref.path.toString + "#" + ref.path.uid
+    override def toString: String =
+      s"${ref.path.toString}#${ref.path.uid} @ $systemUid"
```
just my add hoc format

patriknw commented:
I have updated this. Changed to use the sender node's uid instead of the destination's because that allows to use the serialization optimization when sending same message to several nodes.

This is ready for final review.

johanandren commented on akka-distributed-data/src/test/scala/akka/cluster/ddata/DeltaPropagationSelectorSpec.scala:
```diff
@@ -92,8 +92,8 @@ class DeltaPropagationSelectorSpec extends WordSpec with Matchers with TypeCheck
           false,
-          Map("A" -> Delta(DataEnvelope(deltaA), 1L, 1L), "B" -> Delta(DataEnvelope(deltaB), 1L, 1L)))
-      selector.collectPropagations() should ===(Map(nodes(0) -> expected1, nodes(1) -> expected1))
+          Map("A" -> Delta(DataEnvelope(deltaA), 1L, 1L), "B" → Delta(DataEnvelope(deltaB), 1L, 1L)))
```
These shouldn't be unicode arrows?

johanandren reviewed (approved):
LGTM except for some unicode arrows.

patriknw commented:
this has been running in a jenkins repeat job for several days without failures

chbatey commented on akka-cluster-typed/src/main/scala/akka/cluster/typed/internal/receptionist/ClusterReceptionist.scala:
```diff
@@ -268,34 +311,52 @@ private[typed] object ClusterReceptionist extends ReceptionistBehaviorProvider {
+            Behaviors.stopped
+          } else {
+            // Ok to update from several nodes but more efficient to try to do it from one node.
+            // If cluster.isTerminated we shouldn't try to remove other entries and our own entries should
```
I don't see a `cluster.isTerminated` or is that implied by not being the leader?

patriknw replied:
right, it was a leftover. now it's handled via 
```
case ClusterShuttingDown      => NodeRemoved(setup.cluster.selfUniqueAddress)
```
and that will stop the actor.
That will happen before other MemberRemoved when shutting down.

chbatey reviewed (approved):
LGTM, that was an odd scenario
